RepositoryProvenance: Remove the `originalVcsInfo` property

fviernau opened the pull request:
The purpose of the provenance is to point to some code location. That
location is either a source artifact, a VCS location / revision or
an unknown location.

The main use case in ORT for provenance is to describe what has been
downloaded or scanned. For that purpose the original VCS information is
irrelevant. Remove it to make provenance be usable for that single
purpose. If it should become necessary to also have the original
provenance a separate attribute can be added elsewhere.

Note that finding a scan result for a given package does not involve the
`originalVcsInfo`, in fact all its attribute values are contained also
in `vcsInfo` except for the `resolvedRevision`.

Note: The change is marked as breaking, because "old" scan result storage entries for packages without a revision in their meta-data won't be matched anymore. They would then "automatically" be re-scanned.

sschuberth commented on model/src/main/kotlin/Provenance.kt:
```diff
@@ -64,25 +63,14 @@ data class RepositoryProvenance(
-        // If pkg.vcsProcessed equals originalVcsInfo or vcsInfo this provenance was definitely created when
-        // downloading this package.
-        if (pkg.vcsProcessed == originalVcsInfo || pkg.vcsProcessed == vcsInfo) return true
+        // If this provenance was definitely created when downloading this package.
```
Drop "If".

sschuberth commented on model/src/main/kotlin/ScanResult.kt:
```diff
@@ -54,8 +54,7 @@ data class ScanResult(
             val vcsProvenance = provenance.copy(
-                vcsInfo = provenance.vcsInfo.copy(path = path),
-                originalVcsInfo = provenance.originalVcsInfo?.copy(path = path)
+                vcsInfo = provenance.vcsInfo.copy(path = path)
```
Nit: This now fits in one line.

sschuberth commented:
I believe we need to reword the commit message to provide a more clear and convincing rationale why the `originalVcsInfo` field is not needed anymore. In particular as the field's documentation states:

https://github.com/oss-review-toolkit/ort/blob/e5b135df72f809563296912f99c59149571279a6/model/src/main/kotlin/Provenance.kt#L72-L73

So *how* can we now still match the package in that case?

fviernau commented:
I'll finish this PR after #3991 is in master.

fviernau commented:
> I believe we need to reword the commit message to provide a more clear and convincing rationale why the `originalVcsInfo` field is not needed anymore. In particular as the field's documentation states:
> 
> https://github.com/oss-review-toolkit/ort/blob/e5b135df72f809563296912f99c59149571279a6/model/src/main/kotlin/Provenance.kt#L72-L73
> 
> So _how_ can we now still match the package in that case?

I've added a comment to improve the KDoc for the `originalVcsInfo` property, to make it less mis-leading before removing it.
I've also added a short explanation to the commit message of the commit which removes `originalVcsInfo`. @sschuberth  please see if you like it.

fviernau commented:
@sschuberth @mnonnenmacher this is no more on hold, now ready for review again.

sschuberth commented on model/src/main/kotlin/Provenance.kt:
```diff
@@ -66,11 +66,9 @@ data class RepositoryProvenance(
-     * for the tag again.
+     * The original [VcsInfo] equal to the [meta-data of the package][Package.vcsProcessed], so with an empty resolved
+     * revision (as of 835b197). It is different from [vcsInfo] only in terms of the [resolved revision]
+     * [VcsInfo.resolvedRevision].
```
> 2. All VcsInfo properties except for the resolved revision must be equal

Here, I'm having a problem with the argumentation. If the scan results says to have a scan result for id "foo" of (unresolved) revision "master", then we actually must never use than scan result, as we have no clue whether "master" changed since the time it was scanned.

In other words, shouldn't be exactly the opposite: We do not care about the `revision`s, but the `resolvedRevision`s must match in order to use a stored scan result.

What am I missing here?

sschuberth replied:
Commit message:

> does not involde

"does not involve"

sschuberth commented on downloader/src/funTest/kotlin/BeanUtilsFunTest.kt:
```diff
@@ -72,7 +72,7 @@ class BeanUtilsFunTest : StringSpec() {
                 vcsInfo.type shouldBe VcsType.SUBVERSION
                 vcsInfo.url shouldBe vcsFromCuration.url
-                vcsInfo.revision shouldBe "928490"
+                vcsInfo.revision shouldBe ""
```
Ping typo.

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -283,7 +283,7 @@ class Downloader(private val config: DownloaderConfiguration) {
         )
 
-        return RepositoryProvenance(vcsInfo, pkg.vcsProcessed.takeIf { it != vcsInfo })
+        return RepositoryProvenance(vcsInfo, pkg.vcsProcessed)
```
What would help me to understand better:

- Please explain why "`pkg.vcsProcessed.resolvedRevision` is always blank" is actually correct (because `resolvedRevision` is only populated by the downloader).
- Please explain why "`vcsInfo.resolvedRevision` is never blank" is actually correct.
- In exchange drop the last paragraph.

sschuberth commented on model/src/main/kotlin/Provenance.kt:
```diff
@@ -114,3 +114,6 @@ private class ProvenanceDeserializer : StdDeserializer<Provenance>(Provenance::c
 }
+
+private fun VcsInfo.equalsIgnoreResolvedRevision(other: VcsInfo): Boolean =
+    type == other.type && url == other.url && revision == other.revision && path == other.path
```
Maybe `copy(resolvedRevision = "") == other.copy(resolvedRevision = "")`?

sschuberth commented on model/src/main/kotlin/Provenance.kt:
```diff
@@ -66,11 +66,9 @@ data class RepositoryProvenance(
-     * would be no way to match the package to the [Provenance] without downloading the source code and searching
-     * for the tag again.
+     * The [originalVcsInfo] is is equal to the [meta-data of the package][Package.vcsProcessed]. Thus, it has an empty
+     * resolved revision (as of 835b197). All attributes of [originalVcsInfo], except for the resolved revision, equal
```
Please reword a bit to be able to drop the reference to a concrete commit SHA1. Or maybe just drop the commit completely?

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -276,7 +276,7 @@ class Downloader(private val config: DownloaderConfiguration) {
         )
 
-        return RepositoryProvenance(vcsInfo, pkg.vcsProcessed)
+        return RepositoryProvenance(vcsInfo)
```
> For that purpose the original VCS information is irrelevant.

... because ...

sschuberth reviewed (dismissed):
As discussed by voice.

fviernau commented:
@sschuberth I updated my PR to (hopefully exactly) what we discussed.
